systemtests/tests/test_direct_upload.py

```diff
@@ -0,0 +1,155 @@
+    def test_upload(self, tecken_client: TeckenClient, zip_archive: FakeZipArchive):
+        response = tecken_client.upload(zip_archive.file_name)
+        assert response.status_code == 201
+        zip_archive.uploaded = True
+
+    @small_and_large_archives
+    def test_download(self, tecken_client: TeckenClient, zip_archive: FakeZipArchive):
+        if not zip_archive.uploaded:
+            pytest.skip("upload failed")
+
+        for sym_file in zip_archive.members:
+            response = tecken_client.download(sym_file.key())
+            assert response.status_code == 200
+            [redirect] = response.history
+            assert redirect.status_code == 302
+
+    @small_archive
+    def test_head_request(
+        self, tecken_client: TeckenClient, zip_archive: FakeZipArchive
+    ):
+        if not zip_archive.uploaded:
+            pytest.skip("upload failed")
+
+        for sym_file in zip_archive.members:
+            response = tecken_client.download(sym_file.key(), method="HEAD")
+            assert response.status_code == 200
+            assert not response.history
+
+    @small_archive
+    @pytest.mark.nginx
+    def test_headers(self, tecken_client: TeckenClient, zip_archive: FakeZipArchive):
+        if not zip_archive.uploaded:
+            pytest.skip("upload failed")
+
+        # These are exclusively security headers added by nginx
+        expected_tecken_headers = {
+            "X-Content-Type-Options": "nosniff",
+            "X-Frame-Options": "DENY",
+            "Content-Security-Policy": ANY,
+            "Strict-Transport-Security": "max-age=31536000",
+        }
+        # These are headers sent by the storage backend (S3 or GCS)
+        expected_storage_headers = {
+            "Content-Encoding": "gzip",
+            "Content-Length": ANY,
+        }
+        response = tecken_client.download(zip_archive.members[0].key())
+        [redirect] = response.history
+
+        actual_tecken_headers = {
+            name: redirect.headers[name] for name in expected_tecken_headers
+        }
+        assert actual_tecken_headers == expected_tecken_headers
+
+        actual_storage_headers = {
+            name: response.headers[name] for name in expected_storage_headers
+        }
+        assert actual_storage_headers == expected_storage_headers
+
+    @small_archive
+    def test_code_info_lookup(
+        self, tecken_client: TeckenClient, zip_archive: FakeZipArchive
+    ):
+        if not zip_archive.uploaded:
+            pytest.skip("upload failed")
+
+        for sym_file in zip_archive.members:
+            response = tecken_client.download(
+                sym_file.code_info_key(), allow_redirects=False
+            )
+            assert response.status_code == 302
+            redirect_key = unquote(response.headers["location"])[1:]
+            assert redirect_key == sym_file.key()
+
+
+class TestDirectTryUpload:
+    # Tests within a class scope are executed in definition order.
```

Comment:
I'm not sure why you wrote this. Is this implying that one test relies on another test passing?

Reply:
We talked about this in Zoom. We're going to adjust the limiting for uploads (bug 1910613) and that might reduce the problem we're trying to work around here where we want to minimize the number of uploads as much as possible.